compiler/front/cli_reporter.nim

```diff
@@ -3661,9 +3661,11 @@ proc reportHook*(conf: ConfigRef, r: Report): TErrorHandling =
 
   elif r.kind in rdbgTracerKinds and conf.isDefined(traceDir):
     rotatedTrace(conf, r)
+    conf.m.reports.freeReport(r)
```

Comment:
Report freeing should never happen in the report writing, it should happen in the `options.nim` *after* the report has been considered for writing. A good place to consider putting report freeing is 

https://github.com/nim-works/nimskull/blob/309a57c88b332307b5b40b85194dec2feb6a3d13/compiler/front/options.nim#L516

At this location we still have ID information, therefore all changes with `emptyReport` and `reportId` built into the `Report` object will no longer be necessary.

Reply:
I think the issue is that `handleReport` simply takes the PNode, gets the reportId, like up the report and passes it on -- we never call that proc. Changing this seemed to be rather tricky.

https://github.com/nim-works/nimskull/blob/devel/compiler/front/msgs.nim#L504-L512

Reply:
Reworking it might be less trivial due to some unnecessary interactions, but I would prefer to streamline the reporting pipeline to ensure there is a single point where `id -> report` conversion happens rather than to put cleanup operations on the wrong place, handling earlier design flaws that I made. 

Although the code you provided can also contain a cleanup action, replacing `conf.m.writtenSemReports` with a more general `OrderedTable`-based solution that would remove all reports that had already been written, ensuring both memory reclamation and uniqueness of the write action for each report.